### ISSUE TITLE
--cluster-advertise is not easily automatable

### ISSUE THREAD

**Issue by cpuguy83** [docker/docker#17047]
Trying to use docker-machine to setup a swarm cluster and wanting to utilize multihost networking.
When setting up a machine, I cannot pass in `--cluster-advertise` since I don't know the machine's IP until after it is created.
To get around this, after the machine is setup I have to go in and manually add this flag with the correct IP address and restart the daemon.

Here is my script for doing this: https://gist.github.com/cpuguy83/79ad11aaf8e78c40ca71

What might be nice is if `--cluster-store` is specified but `--cluster-advertise` is not, if we could default to use the first found tcp address from `-H`?
Just a thought.... or provide a cluster management API.

**Comment by cpuguy83** [comment]
Also, the daemon does not complain if `--cluster-advertise` is not specified.

**Comment by mavenugo** [comment]
@cpuguy83 thanks for raising the issue. @aluzzardi & @tombee raised the same valid concern yesterday. Infact, during our initial design, our plan is to automate the --cluster-advertise through some intelligence at the discovery layer. We couldnt make use of -H due to possible use of an internal ip-address as the -H option. the `--discover-advertise` must be a cluster-wide reachable IP and -H doesnt guarantee such a thing. Hence, a possible logic could be 

```
if <user-specified-cluster-advertise> != "" {
    return <user-specified-cluster-advertise>, nil
}
if isExternalIP(cluster-advertise) {
    return find-routable-ip-towards-cluster-advertise, nil
}
return nil, error
```

**Comment by tombee** [comment]
Yup, +1.

When we provision an engine using machine, how would we detect the external facing address? (Usually the address of eth1 I think?)

**Comment by dave-tucker** [comment]
@tombee @mavenugo `wget -qO- icanhazip.com`

**Comment by tombee** [comment]
@dave-tucker, yeah it's just that this won't work for virtualbox etc. right?

```
tom at debian in ~
$ docker-machine ssh default 
 ...
docker@default:~$ wget -qO- icanhazip.com
79.97.x.x
docker@default:~$ ifconfig eth1
eth1      Link encap:Ethernet  HWaddr 08:00:27:6D:D3:82  
          inet addr:192.168.99.100  Bcast:192.168.99.255  Mask:255.255.255.0
          inet6 addr: fe80::a00:27ff:fe6d:d382/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:15 errors:0 dropped:0 overruns:0 frame:0
          TX packets:17 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:2128 (2.0 KiB)  TX bytes:1792 (1.7 KiB)
```

**Comment by mavenugo** [comment]
@tombee @dave-tucker with this PR we are trying to see if we can make `--cluster-advertise` fully optional by automatically detecting the address. Hence I think if the --cluster-store is a remote IP and is reachable via whichever interface in that machine (eth1 or otherwise), then we can select that address as the --cluster-advertise at the docker engine. WDYT ?

**Comment by dave-tucker** [comment]
@mavenugo yeah that seems sensible. pick the address on the interface that is the next hop to `--cluster-store`.

**Comment by tombee** [comment]
Sounds good :)
On Thu 15 Oct 2015 at 12:36 p.m., Dave Tucker <notifications@github.com>
wrote:

> @mavenugo <https://github.com/mavenugo> yeah that seems sensible. pick
> the address on the interface that is the next hop to --cluster-store.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/17047#issuecomment-148360883>.
>
-- 
Tom Barlow
thomas.barlow@docker.com
@tomwbarlow

**Comment by cpuguy83** [comment]
Maybe only if --cluster-advertise is set to auto or something?

I think at least with it no erroring out when it's empty it may be a 1.9 concern, lest we have a bunch people saying overlay isn't working.

**Comment by mavenugo** [comment]
@cpuguy83 maybe if we don't error out, but just Warn the user that we are unable to identify --cluster-advertise, that will take care of it. Wont it ? 
I think it is reasonable to explain/document the fact that discovery is a pre-req for multi-host networking.

**Comment by icecrime** [comment]
I understood all this wrong so yes, I think this needs to be improved for 1.9.0. Now that @mavenugo clarified the situation, let's keep in mind that discovery serves two purposes:

1. Overlay network connectivity
2. API endpoints discovery for Swarm

No. 1 requires the host ip. This one is easy to automate (see above), and doesn't even need to relate in any way to what is specified as part of `--cluster-advertise` as far as I understand.

No. 2 requires a `host:port` combination on which a Docker API endpoint is exposed. This one, from my point of view, would be risky to guess: the best information we have is through `-H`, but who knows if the HTTP API is behind nginx or who knows what.

I expect many people to care about 1, and much fewer to care about 2 (and this group of users is already used to specifying the API endpoint as part of the command line).

Ultimately, I think I agree with the strategy described here https://github.com/docker/docker/issues/17047#issuecomment-148316861, **but** I'd like to confirm one thing: the address that results of this "deduction logic" will only be used internally by libnetwork to form the overlay network, but will not be treated as if the user specified it as part of `--cluster-advertise` (hence it's _not_ going to be written to the kv-store).

**Comment by mrjana** [comment]
@icecrime If you don't write the auto-detected node self IP to the kv-store it will still not work for overlay since we also need to learn atleast one neighbor IP. Imagine a cluster where no daemon added the `--cluster-advertise` option. In that case overlay will simply not work. So the right solution to automate this in discovery layer may be only when `--cluster-advertise=auto` or somthing like that is specified

**Comment by aluzzardi** [comment]
> No. 2 requires a host:port combination on which a Docker API endpoint is exposed. This one, from my point of view, would be risky to guess: the best information we have is through -H, but who knows if the HTTP API is behind nginx or who knows what.

I agree - it's risky.

One point we discussed a while back was to **not** include the port, at least in the key, but have it as a value (e.g. `Key(192.168.0.1) -> Value({APIPort: 2375}) ).

Or maybe even `Key(Unique ID) -> Value({IP: "192.168.0.1", APIPort: 2375})`

Overlay doesn't require the engine to listen on a TCP port and the current store format just assumes so.

**Comment by aluzzardi** [comment]
> the address that results of this "deduction logic" will only be used internally by libnetwork to form the overlay network, but will not be treated as if the user specified it as part of --cluster-advertise (hence it's not going to be written to the kv-store).

Yeah, I would actually do the auto detection and set that as the default flag value for `--cluster-advertise`.

That way `docker -h` would actually show which IP we are trying to use.

WDYT?

**Comment by icecrime** [comment]
Ok, so would that work?

- `--cluster-advertise` has a special `auto` value that does best effort host-ip deduction
- `auto` is the default value for the `--cluster-advertise` flag
- We document this "best effort" strategy and make it clear that Swarm users have an extra manual thing to do, which is... specify the `ip:port` themselves? Do we want `auto:port`? :mask:

**Comment by mrjana** [comment]
@icecrime That works for libnetwork and overlay

**Comment by cpuguy83** [comment]
Swarm isn't using this at all right now is it?

**Comment by abronan** [comment]
@cpuguy83 No no yet as we miss all the non kv related methods to do discovery.

**Comment by aluzzardi** [comment]
@icecrime I'd rather do the detection and stick that a the default value of the flag rather than having a magic `auto` keyword.

As for Swarm ... I don't know. Do we want `--cluster-advertise` to carry both the IP and Port or should it just be the IP? We could maybe dump the `-H` things into the discovery value.

**Comment by dave-tucker** [comment]
:+1: to @aluzzardi's suggestion
As for carrying both the IP and Port in the one parameter, I would split them in to
`--cluster-address` and `--cluster-port` as this allows for maximum flexibility.

- Neither specified = auto discovery for both address and port
- `--cluster-address` = manual address, auto discovered port
- `--cluster-port` = auto address, manual port
- `--cluster-address` and `--cluster-port` = both manually specified

I think only `--cluster-address` would be needed for 1.9, so the details of the port flag can be hashed out later.

**Comment by dave-tucker** [comment]
I guess you could achieve the above with an `auto` keyword too. Flags vs Keyword is a matter of taste - I'm a lover of :flags: though

**Comment by icecrime** [comment]
One thing to keep in mind: specifying `--cluster-advertise` today _enables_ discovery: we have to revise this if make auto detected IP the default.

**Comment by cpuguy83** [comment]
If we could specify the interface to use instead of the actual IP, this would help dramatically since that is fairly automatable.

**Comment by cpuguy83** [comment]
Any news here?

**Comment by moxiegirl** [comment]
+1 to @cpuguy83 questin @mavenugo are we fixing this for 1.9.0 it significantly changes the documentation as it significantly changes the UX. Right now, the workaround is:

```
docker-machine ssh MACHINE 'sudo sh -c "set -ex; /etc/init.d/docker stop || true; sed -i '\''5i     --cluster-advertise='$(docker-machine ip MACHINE)':0\"'\'' 
```

**Comment by mavenugo** [comment]
@cpuguy83 am inclined towards making this simple using your idea of accepting both ip-address of the interface or the interface-name. WDYT ?
I will push a PR to this effect.

**Comment by cpuguy83** [comment]
I'm +1 to that obviously :)
I think when deploying people should always know which interface is the correct one upon machine creation.

**Comment by mavenugo** [comment]
@cpuguy83 the question though would be the expectation on the dynamic nature of ip-address assignment to the interface. What if the ip-address of that interface changed post daemon start ?
I dont think we can tolerate such dynamism.

**Comment by tiborvass** [comment]
Closed by #17364
